### PR TITLE
Return appropriate controller-access for remote users.

### PR DESCRIPTION
**Pull request by howbazaar**
This bug addresses two issues:
  http://pad.lv/1621679 - Login succeeds when permissions forbid it but all requests are denied afterwards
  http://pad.lv/1629089 - API Login ACL response values differ from CLI

The authenticiation code was not adding higher level "everyone" controller access to the login response.
Also, the same area of code would not reject access to a model for an unknown user if there was an "everyone" controller permission.

Tidied up the tests a bit to use test cleanup to close api connections rather than adding a defer call to every test.

### CODE REVIEW

File: apiserver/admin_test.go [hunk 1]
```diff
@@ -804,13 +787,49 @@ func (s *loginSuite) TestOtherEnvironmentWhenNotController(c *gc.C) {
...
-	defer st.Close()
 
 	err := st.Login(machine.Tag(), password, "nonce", nil)
-	c.Assert(errors.Cause(err), gc.DeepEquals, &rpc.RequestError{
-		Message: "permission denied",
-		Code:    "unauthorized access",
+	assertPermissionDenied(c, err)
+}
+
+func (s *loginSuite) loginUserWithAccess(c *gc.C, info *api.Info) (*state.User, params.LoginResult) {
```
Comment (wallyworld) [inline apiserver/admin_test.go]: what does "WithAccess" mean here?
It seems like it should mean "log in this specified user with this specified access" but that doesn't seem to be the case?

File: apiserver/admin.go [hunk 2]
```diff
@@ -187,10 +187,34 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
...
+			// the macaroon was successfully discharged, this means that the user
+			// should not have access to the model, so is permission denied.
+			if !controllerOnlyLogin && permission.ReadAccess.GreaterModelAccessThan(modelAccess) {
+				return fail, errors.Trace(common.ErrPerm)
+			}
+
+			// It is possible that there isn't a controllerUser for the remote user,
+			// so in that situation, we allows the remote user login access.
+			if permission.LoginAccess.GreaterControllerAccessThan(controllerAccess) {
+				controllerAccess = permission.LoginAccess
```
Comment (wallyworld) [inline apiserver/admin.go]: This doesn't look right. And the new tests seems to confirm that. If a permission has been granted to everyone@external, and then that's what will be used for external users (if greater than explicit permission)
There's already a helper for this apiserver/common/maybeUseGroupPermission()

File: apiserver/admin_test.go [hunk 3]
```diff
@@ -804,13 +787,49 @@ func (s *loginSuite) TestOtherEnvironmentWhenNotController(c *gc.C) {
...
+	c.Check(result.UserInfo.Identity, gc.Equals, user.Tag().String())
+	c.Check(result.UserInfo.ControllerAccess, gc.Equals, "login")
+	c.Check(result.UserInfo.ModelAccess, gc.Equals, "admin")
+}
+
+func (s *loginSuite) TestLoginResultModelUserEveryoneCreateOnlyNonLocal(c *gc.C) {
+	info, cleanup := s.setupServer(c)
+	defer cleanup()
+
+	setEveryoneAccess(c, s.State, s.AdminUserTag(c), permission.AddModelAccess)
```
Comment (wallyworld) [inline apiserver/admin_test.go]: Mentioned in another comment, this is the test that appears to check for incorrect behaviour. We are granting addmodel to everyone, but below only checking for "login"

File: apiserver/admin.go (hunk 2 above)
Comment (wallyworld) [inline apiserver/admin.go]: Discussed live: as of now, permissions are expected be defined explicitly, either for the specific user, or via the special "everyone@external" group. So if there is no controller access record for the external user, and there is no "everyone@external" permission to fall back on, then we should return a permission denied. We can loosen that strict approach if needed, but right now it's best to start as restrictive as possible. All the current workflows that I am aware of will work with this strict approach.

**Comment by howbazaar** [comment]
@wallyworld updated the behaviour, extracted into its own function.

File: apiserver/admin_test.go [hunk 4]
```diff
@@ -1049,12 +1065,24 @@ func (s *macaroonLoginSuite) TestRemoteUserLoginToModelWithExplicitAccess(c *gc.
 	c.Check(err, jc.ErrorIsNil)
 	c.Assert(result.UserInfo, gc.NotNil)
 	c.Check(result.UserInfo.Identity, gc.Equals, remoteUserTag.String())
-	c.Check(result.UserInfo.ControllerAccess, gc.Equals, "login")
+	c.Check(result.UserInfo.ControllerAccess, gc.Equals, "addmodel")
 	c.Check(result.UserInfo.ModelAccess, gc.Equals, "write")
 }
 
+func (s *macaroonLoginSuite) AddControllerUser(c *gc.C, user names.UserTag, access permission.Access) {
```
Comment (wallyworld) [inline apiserver/admin_test.go]: prefer lower case addControllerUser()

**Comment by howbazaar** [comment]
$$merge$$

**Comment by howbazaar** [comment]
$$merge$$

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**